module/zfs/metaslab.c

```diff
@@ -1171,13 +1204,19 @@ metaslab_cf_alloc(metaslab_t *msp, uint64_t size)
 
 	if ((*cursor + size) > *cursor_end) {
 		range_seg_t *rs;
-
-		rs = avl_last(&msp->ms_size_tree);
-		if (rs == NULL || (rs->rs_end - rs->rs_start) < size)
+		for (rs = avl_last(&msp->ms_size_tree);
+		    rs != NULL && rs->rs_end - rs->rs_start >= size;
+		    rs = AVL_PREV(&msp->ms_size_tree, rs)) {
+			*cursor = rs->rs_start;
+			*cursor_end = rs->rs_end;
+			if (!metaslab_check_trim_conflict(msp, cursor, size,
+			    1, *cursor_end)) {
+				/* segment appears to be acceptable */
```

Comment:
The purpose to place this into each allocator is to have a higher chance of choosing a suitable block. If a block is unsuitable, the allocator can keep searching according to its own algorithm design. If it were placed higher up the stack (say, in metaslab_block_alloc), we are getting fairly constrained as to the decisions we can make - it pretty much boils down to a go-nogo at that point. If an allocator chooses an unsuitable block (because it wasn't aware of trim), we might just have to flatly reject the entire metaslab.